torch/tensor.py

```diff
@@ -728,4 +728,28 @@ def _update_names(self, names, inplace):
         else:
             return super(Tensor, self).rename(names)
 
+    @property
+    def grad(self):
+        """
+        This attribute is ``None`` by default and becomes a Tensor the first time a call to
+        :func:`backward` computes gradients for ``self``.
+        The attribute will then contain the gradients computed and future calls to
+        :func:`backward` will accumulate (add) gradients into it.
+        """
+        if self.requires_grad and not hasattr(self, "retains_grad") and not self.is_leaf and self._grad is None:
+            warnings.warn("The .grad attribute of a Tensor that is not a leaf Tensor is being accessed. Its .grad "
+                          "attribute won't be populated during autograd.backward(). If you indeed want the gradient "
+                          "for a non-leaf Tensor, use .retain_grad() on the non-leaf Tensor. If you access the "
```

Comment:
just some nits (CC @vincentqb):

- "if you access" -> "if you accessed"

- "the non-leaf Tensor by mistake, make sure you access the leaf Tensor instead"

There's not a general "the leaf Tensor" to access.  This sentence isn't really actionable, you would have to know _why_ they accessed to the non-leaf tensor to know, so I'd just get rid of it.

- We should link to documentation instead of pull requests.